### ISSUE TITLE
Some OAuth flows deprecated soon

### ISSUE THREAD

**Issue by rrech** [Ylianst/MeshCentral#4450]
Hi Ylian!

Some OAuth flows will be deprecated soon. More information here:
https://developers.googleblog.com/2022/02/making-oauth-flows-safer.html

At this moment, to use Google Drive backup resource, if one sets and use a new credential to fits new Google criteria , returns an error 400 "invalid_request". The error description could be found here: https://support.google.com/accounts/answer/12379384?hl=en

I don't know if this was addressed yet.

Thanks in advance!

**Comment by Ylianst** [comment]
So, some OAuth flows will be deprecated, specifically loopback and out-of-band. I don't think MeshCentral uses these, so we should be fine. Regardless of these removals, are you saying the Google Drive auto-backup feature does not currently work?

**Comment by rrech** [comment]
Sorry about the title, it was implicit that are "some". I am not a native English speaker, and I ask you to kindly take this into account.

For now, it is working fine.

I've received a communication from Google indicating that the credential that I use in MeshCentral will be deprecated. Setted a new one, when I try to use on MeshCentral, returns an error informing that is a invalid request.

If I misunderstood the situation, I apologize.

**Comment by rrech** [comment]
Error message when setting a new credential is:

--------------------------------------------------
Error 400: invalid_request

You can't sign in to this app because it doesn't comply with Google's OAuth 2.0 policy for keeping apps secure.

You can let the app developer know that this app doesn't comply with one or more Google validation rules.

Request details:

The content in this section has been provided by the app developer. This content has not been reviewed or verified by Google.

If you’re the app developer, make sure that these request details comply with Google policies.

redirect_uri: urn:ietf:wg:oauth:2.0:oob
--------------------------------------------------

**Comment by soapster** [comment]
+1 for this.

**Comment by soapster** [comment]
Same for me. My drive backup stopped a few days ago. Today, I went to re-connect it...

Error 400: invalid_request

You can't sign in to this app because it doesn't comply with Google's OAuth 2.0 policy for keeping apps secure.

You can let the app developer know that this app doesn't comply with one or more Google validation rules.
[Learn more](https://support.google.com/accounts/answer/12379384)
Request Details
The content in this section has been provided by the app developer. This content has not been reviewed or verified by Google.
If you’re the app developer, make sure that these request details comply with Google policies.
redirect_uri: urn:ietf:wg:oauth:2.0:oob

**Comment by soapster** [comment]
Something to note. If I put the project back to testing on the Google console, the process works.  But this will make it so that it will fail in a number of days.

**Comment by omriasta** [comment]
Yep, this only works for me for about 5-6 days and then stops backups again. This is related to https://github.com/Ylianst/MeshCentral/issues/4358
I wish I knew anything about NodeJS, I would be happy to look at the code but I really don't feel I could help much.

**Comment by soapster** [comment]
@Ylianst This is still an issue. This only works for a limited time while the project on Google is in testing. If I publish the app on the console, the google Drive integration fails with the following error.
![22-30 -42](https://user-images.githubusercontent.com/8184435/198176881-7673a8c5-4126-4669-a1a3-9e5543b5ea0e.jpg)

The strange thing is, this works fine on my setup at work. But I cannot figure out what is different between there and home. At work, we are using a free gmail account, just like I am at home.   Maybe you could do a video on the full setup, start to finish on how to make this work.

**Comment by soapster** [comment]
This is the error details.

![22-32 -42](https://user-images.githubusercontent.com/8184435/198177149-68f14912-cd5c-40e2-9c77-318009d23f93.jpg)

**Comment by zikeji** [comment]
A temporary solution is to edit `node_modules/meshcentral/mesh_user.js`, line 4592, and replace `urn:ietf:wg:oauth:2.0:oob` with `http://localhost:1/`, then add that same URL to your redirect URIs in Google's API Console.

![image](https://user-images.githubusercontent.com/9063612/218389343-45e8fe34-b1f6-40b8-9cca-4ddfac5be949.png)

You can then go through the normal steps, and when you get redirected you'll find the code in the URL which you'll copy. You'll get redirected to `http://localhost:1/?code=CODE&scopes=SCOPES`. Just copy the code, not the scopes part, and supply that code.

Then go to the console and run `autobackup` to test if. If it succeeds, you're good to go! You can revert the change, or keep it. I reverted it and autobackups functioned without issue, so it doesn't need to remain.

**Comment by omriasta** [comment]
Actually, I think it now works for me, the issue is that the app must be in production AND use the above hack from @zikeji

**Comment by zikeji** [comment]
Ah, yes, a Google API project that is in testing mode automatically invalidates each token after a week since it's for "testing". I also just found that out the hard way in another project.

**Comment by omriasta** [comment]
yep, and when you create credentials in the google cloud console, make sure you select web application as the type.....for anyone else looking on here. Hopefully the dev can write this in so that meshcentral can just capture the code....I have just never written anything for node....

**Comment by omriasta** [comment]
If the "app" is set to testing, it will work as it is right now but the token expires every few days and needs to be setup again. If the "app" is set to production, the current setup give the oob error above. Modifying as per @zikeji above will work in production but it requires to manually copy the code from the browser to the mesh central interface.

**Comment by zikeji** [comment]
> I can look into this next week, So does the Google drive not work at all? Are there certain steps we have to follow for it to work currently?

I just updated to 1.1.11 and reset my configuration to double check / update my instructions here https://github.com/Ylianst/MeshCentral/issues/4450#issuecomment-1427431575

The reason it fails is because MeshCentral uses a removed feature that would get the oauth screen to display the code afterwards, now you must redirect. My workaround changes the redirect URI in the code (which is necessary vs. just altering the URL) and then you copy the URL from the GET parameters.

In terms of fixing this in MeshCentral itself, the solution would be to remove the "Code" entry step of the little wizard and instead add a path to express that handles the redirect from Google, and updating the instructions in the wizard to have you input the proper redirect URI into Google Cloud. e.g. `https://mesh.domain.com/oauth/callback`. Then that path just reads code from the GET params and saves that to the database.

**Comment by si458** [comment]
Thank you @zikeji for the very helpful hints!

Similar to how we do the oauth of logins and sso! Got u!

**Comment by omriasta** [comment]
And after that we can update the documentation on how to setup everything in the Google cloud console.

**Comment by soapster** [comment]
> # Workaround

Thank you @zikeji  I found the instructions clear, and I was able to implement the changes. Working so far. We'll see in a couple of weeks!

**Comment by soapster** [comment]
Well, I haven't touched this since my last comment. It's till working.
![image](https://github.com/Ylianst/MeshCentral/assets/8184435/1178d37f-f3f0-46ed-a277-5df34d186947)

**Comment by alasdaircs** [comment]
+1 : @zikeji workaround seems good. Clearly a proper web flow with working return URI is required. I would give it a go if it were C# but I'm not a Node dev...

The formal docs for the deprecation of OOB flow are here: [Out-Of-Band (OOB) flow Migration Guide](https://developers.google.com/identity/protocols/oauth2/resources/oob-migration)

**Comment by zikeji** [comment]
Naughty automation closing this out

**Comment by si458** [comment]
is this still an issue?
which flows are deprecated?
how can we replicate the issue?
what logs do you see and where?

**Comment by alasdaircs** [comment]
I linked the Out Of Band flow deprecation 11 months ago. If it's not fixed, then it's still broken.

**Comment by zikeji** [comment]
Sir automation, this is not ready to close out.

**Comment by alasdaircs** [comment]
@zikeji , guys, can someone please re-open this issue as it is still a pain-point and re-applying the code fix after each update isn't working for me.

**Comment by si458** [comment]
@alasdaircs done!
if you have to apply patches to your server after each release
please can you maybe submit a PR for us to look at or even merge?
then you wont have to do it in the future!

**Comment by alasdaircs** [comment]
@si458 I only need to do this when I have to re-authenticate the Google Drive backup, which means I have to make the changes detailed in https://github.com/Ylianst/MeshCentral/issues/4450#issuecomment-1427431575, however the real fix would be to implement https://github.com/Ylianst/MeshCentral/issues/4450#issuecomment-1720080680. I'm temped to have a go but as I'm a mainly a C# guy and the source isn't even TS it's going to be an effort for me.

**Comment by si458** [comment]
was this still an issue sorry?
somebody wanted oauth2 for office365 for smtp, so might look into this as well for google
would need to work out all the scopes etc for google drive and backups etc

**Comment by alasdaircs** [comment]
> [@zikeji](https://github.com/zikeji) , guys, can someone please re-open this issue as it is still a pain-point and re-applying the code fix after each update isn't working for me.

And again. Maybe I'll have a go at this with Claude!
